dips/src/lib.rs

```diff
@@ -0,0 +1,388 @@
+// Copyright 2023-, Edge & Node, GraphOps, and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::str::FromStr;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+pub use alloy;
+pub use alloy_rlp;
+
+use alloy::core::primitives::Address;
+use alloy::rlp::{RlpDecodable, RlpEncodable};
+use alloy::signers::{Signature, SignerSync};
+use alloy_rlp::Encodable;
+use thegraph_core::alloy_sol_types::{sol, Eip712Domain, SolStruct};
+use thiserror::Error;
+
+sol! {
+    // EIP712 encoded bytes, ABI - ethers
+    #[derive(Debug, RlpEncodable, RlpDecodable, PartialEq)]
+    struct SignedIndexingAgreementVoucher {
+        IndexingAgreementVoucher voucher;
+        bytes signature;
+    }
+
+    #[derive(Debug, RlpEncodable, RlpDecodable, PartialEq)]
+    struct IndexingAgreementVoucher {
+      // should coincide with signer
+        address payer;
+        // should coincide with indexer
+        address payee;
+         // data service that will initiate payment collection
+        address service;
+        // initial indexing amount max
+        uint256 maxInitialAmount;
+        uint256 maxOngoingAmountPerEpoch;
+        // time to accept the agreement, intended to be on the order
+        // of hours or mins
+        uint64 deadline;
+        uint32 maxEpochsPerCollection;
+        uint32 minEpochsPerCollection;
+        // after which the agreement is complete
+        uint32 durationEpochs;
+        bytes metadata;
+    }
+
+    // the vouchers are generic to each data service, in the case of subgraphs this is an ABI-encoded SubgraphIndexingVoucherMetadata
+    #[derive(Debug, RlpEncodable, RlpDecodable, PartialEq)]
+    struct SubgraphIndexingVoucherMetadata {
+        uint256 pricePerBlock; // wei GRT
+        bytes32 protocolNetwork; // eip199:1 format
+        // differentiate based on indexed chain
+        bytes32 chainId; // eip199:1 format
+        string deployment_ipfs_hash;
+    }
+
+    #[derive(Debug, RlpEncodable, RlpDecodable, PartialEq)]
+    struct SignedCancellationRequest {
+        CancellationRequest request;
+        bytes signature;
+    }
+
+    #[derive(Debug, RlpEncodable, RlpDecodable, PartialEq)]
+    struct CancellationRequest {
+        // should coincide with signer.
+        address payer;
+        // should coincide with indexer.
+        address payee;
+        // data service that will initiate payment collection.
+        address service;
+        // signer of the cancellation, can be signed by either party.
+        address cancellled_by;
+        // should only be usable within a limited period of time.
+        uint64 timestamp;
```

Comment:
I have followed the thread, and it is used in the cancellation request validation. What's the goal of this timestamp?

Reply:
essentially to prevent re-use or any variation of that.